design_drafts/controller_execution_management.md

```diff
@@ -0,0 +1,125 @@
+One requirement we'd like to see is a more flexible controller chaining.
+Controller chaining still implies a fixed configuration in which order the controllers are being executed, however with shared resources.
+A typical use case for this is a safety `JointLimitsController` which is being linked after a `JointEffortController` to enforce an operation within appropriate min/max torque values.
+Another use case could be to clamp a `DiffDriveController` velocity values to optimize for battery lifetime or top acceleration.
+
+In order to realize such a controller chaining, the output of one's controller must be linked as the input to its consecutive controller.
+This further motivates a relaxation that resources such as the `JointHandle` can only be uniquely claimed.
+
+### Sequential Controller Container
+
+Instead of loading each controller individually as an independent controller, we propose a parent container which holds and connects controllers which are configured in a specific order.
+The container has a single input and a single output which is being scoped within the container.
+That is, all controllers loaded within a container don't see any input/output states outside the container.
+The input and output states can then be safely re-routed and linked between individual controllers.
+After all controllers are executed, the final output state is being returned from the container.
+
+The image below shows how this execution model could look like:
+![sequential_execution](images/sequential_execution.png "Sequential Execution Container")
+
+## Parallel Execution of Controllers
+
+Opposite to the sequential order, there are use cases where it makes sense to operate controllers in a parallel fashion.
+`Controller_1`, `Controller_2`, …, `Controller_3` are being executed in parallel/isolated and their output state is being fused with a specific function.
+
+### Separation of Tasks
+
+Having controllers being run in parallel allows a better separation of tasks per controller.
+One easy to grasp example (also if not the most sensible) is a simple `PID-Controller`.
+We could break down each gain (`P`, `I`, `D`) in its own controller and let them calculate in isolation.
+By being able to execute multiple controllers in parallel, the calculation of the actual output can then be to simply add the `P`, `I`, `D` outputs - according to the control law of `p * (x_d - x) + p * (x_d - x) / dt  + i * sum(x_d -x) * dt`.
+A second use case could be again a differential drive controller which computes rotational and translational velocities and a second controller calculates dynamic friction values which then again can be fused together.
+
+### Parallel Controller Container
+
+Just like in the sequential execution, we propose having a container structure which holds a set of loaded controllers which are being executed in parallel and primarily independent from each other.
+When all controllers were executed, each individual calculated values are being joined.
```

Comment:
If the parallel controllers are operating at different frequencies, do we need to be concerned about the rate at which the container is operating at?

Reply:
very good point. I have thought about this, but couldn't really come up with a good solution. We could consider having each controller adhering to the container's frequency, which would make it cumbersome to cluster controllers with different frequencies. 
We could alternatively giving each container it's own instance of an executor which handles the frequencies. That would lead to what I'd call "nested frequencies" where the absolute frequency specified for a controller depends on the container's base frequency. 
Yet an alternative approach could be to update each controller in its desired frequency and only run the "fuse" operation at the container's frequency.

I have to think more about all these approaches and their caveats. Any ideas are merely appreciated.

Reply:
This is starting to sound similar to the QoS policy design with rosbag2 but with frequencies :)

> ...giving each container it's own instance of an executor which handles the frequencies.

I'm in favor of this approach but I don't know of any repercussions off the top of my head. Maybe added complexity? It does need some thought....
I'm not familiar with what you mean by fusing though.

Reply:
Perhaps in the case of a parallel cluster of controllers, variable update rates can be supported by setting a fixed frequency to the fuse operation, a sort of heartbeat.